### ISSUE TITLE
Integrate RegExodus

### ISSUE THREAD

**Issue by soywiz** [jtransc/jtransc#48]
Since Dalvik RT relies on native C++ code for regular expressions, we need an opensource pure java regexp library compatible with Pattern/Match Java APIs.

@tommyettinger suggested me to use RegExodus, that can be found here: https://github.com/tommyettinger/RegExodus

**Comment by tommyettinger** [comment]
That's me! RegExodus is currently in the middle of some difficult bug-fixing to fix various problems exposed by running the compatibility package (`regexodus.regex` is meant to be near-equivalent to `java.util.regex`) through OpenJDK 7's tests for their regex package. Possessive modifiers are definitely not supported (`a++` or `c*+`), character class intersection/union is not supported (because it's insane and bug-prone, plus almost no language implements them, for good reasons), and comments in regexes with whitespace mode, like `(?x) A # this is a comment, this regex matches the letter A` aren't implemented. Entering case-insensitive mode in the middle of a regex with `(?i)blah` isn't working yet because of a bug, but it should work soon. Supplementary characters (above `'\uFFFF'`, unfortunately including most emoji) are not supported, because almost all of the code is designed around char values instead of int.

The current code is stuck on a test where it tries and fails to compile the regex `(?:\w{2,}?)+;` and I'm not exactly sure why it fails (it's in a section of complicated branching/repeating code, which was taken from JRegex and hasn't needed modification until now). Then again, all I can say about that particular regex is that it should probably match an empty string, or 2 or more word chars, but never 1 word char, and I have no idea when that would ever be wanted. It could also be rewritten more simply as `\w{2,}?`.

When I get most of the tests to pass (the ones that are supported or should work but currently don't), I'll release another version of RegExodus, I think `0.2` It's very possible the current master version will work in the meanwhile, since it has the `regexodus.regex` compatibility package that almost duplicates the API of `java.util.regex`, other than some obscure features like bounds transparency which throw `UnsupportedOperationException`s.

**Comment by soywiz** [comment]
No problem. I'm going to integrate it today. It is better that nothing :) And when you fix the stuff I will update it.
I tried mostly to avoid submodules here. Can I copy the repo to a folder (linking to the original repo)? Or should should include a git submodule pointing to your repo?

**Comment by soywiz** [comment]
I have integrated it already, but I'm waiting for your review and to decide how to include it.

```
#48 Integrate RegExodus
- I have added regexodus to jtransc-rt, included the sourcecode in this repository (waiting if the author prefers a git submodule)
- I have added LICENSE, README.md and a link to the original repository
- I have removed GwtIncompatible annotation
- I have removed the regex package, which is integrated directly in java.util.regex

Matcher have some missing functions.

int start(String) and int end(String)
+ some uncommon stuff
```

**Comment by soywiz** [comment]
I also think that a tightly integration could allow to use the Category class in some String and Character cases.
Also changing classes a bit, could totally avoid having to transform flags between one and other.

**Comment by soywiz** [comment]
Ok. So a small update on this. I have changed that the String class doesn't depend on regexodus, to avoid including that many classes in simple classes not using stuff. In that case javascript regexp are used. For String.split and regexp related APIs embedded in String.

Also I plan to release a version in a couple of days (I have to update it with the lastest changes). You are OK with keeping a copy here to avoid modules? or want a git submodule linking to your repo?

**Comment by tommyettinger** [comment]
Oh yeah, submodules can be a hassle. It probably works best if you need to modify it for integration purposes to keep a copy. This sounds like a good approach, what you have.

**Comment by soywiz** [comment]
Cool. I have updated it to the lastest version. That will be included in the 0.3.0 release.
Please, let me know about new updates so I include them in new jtransc releases :)

Closing this.
